File: solutions/cpp/0-1_knapsack.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <utility>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::max;
using std::pair;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

vector<int> RandVector(int len) {
  vector<int> ret;
  default_random_engine gen((random_device())());
  while (len--) {
    uniform_int_distribution<int> dis(0, 99);
    ret.emplace_back(dis(gen));
  }
  return ret;
}

// @include
struct Item {
  int weight, value;
};

int Knapsack(int w, const vector<Item>& items) {
  vector<int> V(w + 1, 0);
  for (int i = 0; i < items.size(); ++i) {
    for (int j = w; j >= items[i].weight; --j) {
      V[j] = max(V[j], V[j - items[i].weight] + items[i].value);
    }
  }
  return V[w];
}
// @exclude

void SmallTest() {
  // The example in the book.
  vector<Item> items = {{20, 65},
                        {8, 35},
                        {60, 245},
                        {55, 195},
                        {40, 65},
                        {70, 150},
                        {85, 275},
                        {25, 155},
                        {30, 120},
                        {65, 320},
                        {75, 75},
                        {10, 40},
                        {95, 200},
                        {50, 100},
                        {40, 220},
                        {10, 99}};
  assert(695 == Knapsack(130, items));
}

int main(int argc, char* argv[]) {
  SmallTest();
  default_random_engine gen((random_device())());
  vector<int> weight, value;
  int n, W;
  if (argc == 1) {
    uniform_int_distribution<int> n_dis(1, 100);
    n = n_dis(gen);
    uniform_int_distribution<int> W_dis(1, 1000);
    W = W_dis(gen);
    weight = RandVector(n), value = RandVector(n);
  } else if (argc == 2) {
    n = atoi(argv[1]);
    uniform_int_distribution<int> W_dis(1, 1000);
    W = W_dis(gen);
    weight = RandVector(n), value = RandVector(n);
  } else {
    n = atoi(argv[1]);
    W = atoi(argv[2]);
    for (int i = 0; i < n; ++i) {
      weight.emplace_back(atoi(argv[3 + i]));
    }
    for (int i = 0; i < n; ++i) {
      value.emplace_back(atoi(argv[3 + i + n]));
    }
  }
  cout << "Weight: ";
  for (int i = 0; i < n; ++i) {
    cout << weight[i] << ' ';
  }
  cout << endl << "Value: ";
  for (int i = 0; i < n; ++i) {
    cout << value[i] << ' ';
  }
  cout << endl;
  vector<Item> items;
  for (int i = 0; i < weight.size(); ++i) {
    items.emplace_back(Item{weight[i], value[i]});
  }
  cout << "Knapsack size = " << W << endl;
  cout << "all value = " << Knapsack(W, items) << endl;
  return 0;
}

File: solutions/python/0_1_knapsack.py
import sys
import random
import collections


def rand_vector(length):
    return [random.randrange(100) for _ in range(length)]


# @include
Item = collections.namedtuple('Item', ('weight', 'value'))


def optimum_subjec_to_capacity(items, capacity):
    # Returns the optimum value when we choose from items[:k + 1] and have a
    # capacity of available_capacity.
    def optimum_subject_to_item_and_capacity(k, available_capacity):
        if k < 0:
            # No items can be chosen.
            return 0

        if V[k][available_capacity] == -1:
            without_curr_item = optimum_subject_to_item_and_capacity(
                k - 1, available_capacity)
            with_curr_item = (0 if available_capacity < items[k].weight else (
                items[k].value + optimum_subject_to_item_and_capacity(
                    k - 1, available_capacity - items[k].weight)))
            V[k][available_capacity] = max(without_curr_item, with_curr_item)
        return V[k][available_capacity]

    # V[i][j] holds the optimum value when we choose from items[:i + 1] and have
    # a capacity of j.
    V = [[-1] * (capacity + 1) for _ in items]
    return optimum_subject_to_item_and_capacity(len(items) - 1, capacity)
# @exclude


def small_test():
    # The example in the book.
    items = [
        Item(w, v)
        for w, v in ((20, 65), (8, 35), (60, 245), (55, 195), (40, 65), (
            70, 150), (85, 275), (25, 155), (30, 120), (65, 320), (75, 75), (
                10, 40), (95, 200), (50, 100), (40, 220), (10, 99))
    ]
    assert 695 == optimum_subjec_to_capacity(items, 130)

    items = (Item(5, 60), Item(3, 50), Item(4, 70), Item(2, 30))
    assert 80 == optimum_subjec_to_capacity(items, 5)


def main():
    small_test()
    if len(sys.argv) == 1:
        n = random.randint(1, 100)
        W = random.randint(1, 1000)
        weight = rand_vector(n)
        value = rand_vector(n)
    elif len(sys.argv) == 2:
        n = int(sys.argv[1])
        W = random.randint(1, 1000)
        weight = rand_vector(n)
        value = rand_vector(n)
    else:
        n = int(sys.argv[1])
        W = int(sys.argv[2])
        weight = [int(sys.argv[3 + i]) for i in range(n)]
        value = [int(sys.argv[3 + i + n]) for i in range(n)]
    print('weight:', *weight)
    print('value:', *value)
    items = [Item(w, v) for w, v in zip(weight, value)]
    print('knapsack size =', W)
    print('all value =', optimum_subjec_to_capacity(items, W))


if __name__ == '__main__':
    main()
